beacon-chain/sync/service.go

```diff
@@ -0,0 +1,134 @@
+//     *  Filter redundant data and unwanted data
+//     *  Drop peers that send invalid data
+//     *  Trottle incoming requests
+type Service struct {
+	ctx            context.Context
+	cancel         context.CancelFunc
+	networkService NetworkService
+	chainService   ChainService
+	hashBuf        chan hash.Hash
+	blockBuf       chan *types.Block
+}
+
+// Config allows the channel's buffer sizes to be changed
+type Config struct {
+	HashBufferSize int
+	BlockBufferSize int
+}
+
+// DefaultConfig provides the default configuration for a sync service
+func DefaultConfig() Config {
+	return Config { 100, 100 }
+}
+
+// NetworkService is the interface for the p2p network.
+type NetworkService interface {
+	BroadcastBlockHash(hash.Hash) error
+	BroadcastBlock(*types.Block) error
+	RequestBlock(hash.Hash) error
+}
+
+// ChainService is the interface for the local beacon chain.
+type ChainService interface {
+	ProcessBlock(*types.Block) error
+	ContainsBlock(hash.Hash) bool
+}
+
+// NewSyncService accepts a context and returns a new Service.
+func NewSyncService(ctx context.Context, cfg Config) *Service {
+	ctx, cancel := context.WithCancel(ctx)
+	return &Service{
+		ctx:      ctx,
+		cancel:   cancel,
+		hashBuf:  make(chan hash.Hash, cfg.HashBufferSize),
+		blockBuf: make(chan *types.Block, cfg.BlockBufferSize),
+	}
+}
+
+// SetNetworkService sets a concrete value for the p2p layer.
+func (ss *Service) SetNetworkService(ps NetworkService) {
+	ss.networkService = ps
+}
+
+// SetChainService sets a concrete value for the local beacon chain.
+func (ss *Service) SetChainService(cs ChainService) {
+	ss.chainService = cs
+}
+
+// Start begins the block processing goroutine.
+func (ss *Service) Start() {
+	log.Info("Starting service")
+	go run(ss.ctx.Done(), ss.hashBuf, ss.blockBuf, ss.networkService, ss.chainService)
+}
+
+// Stop kills the block processing goroutine, but does not wait until the goroutine exits.
+func (ss *Service) Stop() error {
+	log.Info("Stopping service")
+	ss.cancel()
+	return nil
+}
+
+// ProcessBlockHash accepts a block hash.
+// New hashes are forwarded to other peers in the network (unimplemented), and
+// the contents of the block are requested if the local chain doesn't have the block.
+func (ss *Service) ProcessBlockHash(h hash.Hash) {
+	if ss.chainService.ContainsBlock(h) {
+		return
+	}
+
+	ss.hashBuf <- h
+	ss.networkService.BroadcastBlockHash(h)
+}
+
+// ProcessBlock accepts a block to potentially be included in the local chain.
+// The service will filter blocks that have not been requested (unimplemented).
+func (ss *Service) ProcessBlock(b *types.Block) error {
+	h, err := b.Hash()
+	if err != nil {
+		return err
+	}
+
+	if ss.chainService.ContainsBlock(h) {
+		return nil
+	}
+
+	ss.blockBuf <- b
+	ss.networkService.BroadcastBlock(b)
+
+	return nil
+}
+
+func run(done <-chan struct{}, hashBuf <-chan hash.Hash, blockBuf <-chan *types.Block, ps NetworkService, cs ChainService) {
```

Comment:
rename to something more aligned with what the goroutine does

Reply:
i gave it a generic name because the intention was for there to be only one main goroutine. i think services can benefit from restricting itself to one goroutine by making the logic more concise. it'd be nice to see this more often in our code base. thoughts?

Reply:
to be more specific, multiple goroutines in a single service means sharing state across multiple goroutines, and potential race conditions. 

restricting to a single goroutine could also make service management simpler by replacing `Start` with `go service.Run()` and `Stop` to `ctx.cancel()`.

Reply:
Having a run function as a main one is a good solution though, im just concerned itll grow too much given how these services encompass so much. For example, the chainservice could become extremely complex down the line and it could be tough to have everything in a single routine.